src/librawspeed/decoders/ArwDecoder.cpp

```diff
@@ -270,6 +277,79 @@ void ArwDecoder::DecodeUncompressed(const TiffIFD* raw) const {
   }
 }
 
+void ArwDecoder::DecodeLJpeg(const TiffIFD* raw) const {
+  uint32_t width = raw->getEntry(TiffTag::IMAGEWIDTH)->getU32();
+  uint32_t height = raw->getEntry(TiffTag::IMAGELENGTH)->getU32();
+  uint32_t bitPerPixel = raw->getEntry(TiffTag::BITSPERSAMPLE)->getU32();
+
+  switch (bitPerPixel) {
+  case 8:
+  case 12:
+  case 14:
+    break;
+  default:
+    ThrowRDE("Unexpected bits per pixel: %u", bitPerPixel);
+  }
+
+  if (width == 0 || height == 0 || height % 2 != 0 || width > 9728 ||
+      height > 6656)
+    ThrowRDE("Unexpected image dimensions found: (%u; %u)", width, height);
+
+  mRaw->dim = iPoint2D(width, height);
+
+  const uint32_t tilew = raw->getEntry(TiffTag::TILEWIDTH)->getU32();
+  const uint32_t tileh = raw->getEntry(TiffTag::TILELENGTH)->getU32();
+
+  if (!(tilew > 0 && tileh > 0))
+    ThrowRDE("Invalid tile size: (%u, %u)", tilew, tileh);
+
+  assert(tilew > 0);
+  const uint32_t tilesX = roundUpDivision(mRaw->dim.x, tilew);
+  if (!tilesX)
+    ThrowRDE("Zero tiles horizontally");
+
+  assert(tileh > 0);
+  const uint32_t tilesY = roundUpDivision(mRaw->dim.y, tileh);
+  if (!tilesY)
+    ThrowRDE("Zero tiles vertically");
+
+  const TiffEntry* offsets = raw->getEntry(TiffTag::TILEOFFSETS);
+  const TiffEntry* counts = raw->getEntry(TiffTag::TILEBYTECOUNTS);
+  if (offsets->count != counts->count) {
+    ThrowRDE("Tile count mismatch: offsets:%u count:%u", offsets->count,
+             counts->count);
+  }
+
+  // tilesX * tilesY may overflow, but division is fine, so let's do that.
+  if ((offsets->count / tilesX != tilesY || (offsets->count % tilesX != 0)) ||
+      (offsets->count / tilesY != tilesX || (offsets->count % tilesY != 0))) {
+    ThrowRDE("Tile X/Y count mismatch: total:%u X:%u, Y:%u", offsets->count,
+             tilesX, tilesY);
+  }
+
+  mRaw->createData();
+  #ifdef HAVE_OPENMP
+  #pragma omp for schedule(static)
+  #endif
+    for (uint32_t tile = 0U; tile < offsets->count; tile++) {
+      const uint32_t tileX = tile % tilesX;
+      const uint32_t tileY = tile / tilesX;
+      const uint32_t offset = offsets->getU32(tile);
+      const uint32_t length = counts->getU32(tile);
+
+      LJpegDecoder decoder(ByteStream(DataBuffer(mFile.getSubView(offset, length),
+                                                Endianness::little)),
+                          mRaw, true);
+      decoder.decode(tileX * tilew, tileY * tileh, tilew, tileh, false);
+    }
+
+  const TiffEntry* origin_entry = raw->getEntry(TiffTag::DEFAULTCROPORIGIN);
+  const TiffEntry* size_entry = raw->getEntry(TiffTag::DEFAULTCROPSIZE);
```

Comment:
As I already mentioned, [let's not use these but the dimensions in tag 0x7038](https://github.com/darktable-org/rawspeed/pull/386#discussion_r1154150298). That'll give us the same active area size as uncompressed ARW, and consequently, the same crop w/ no need for cameras.xml adjustments.

Reply:
Seems to work, I just pushed the changes!

> You could even go a step further and use those directly as width and height on L281-L282 without needing to do anything extra here. The LJpeg decoder should take of the padding just fine.

This however does not work, as the width & height need to be multiples of the tile size to make the tiling code work, or how did you mean that?

Reply:
They don't. The tiling code rounds up to the required multiple.

Reply:
> We then need a new mode that'll tell us to ignore any further cameras.xml crop.

Can we make the mandatory cropping step according to the values in ``cameras.xml`` optional by just skipping it on a decoder basis? 
But I guess it would be cleaner and more flexible to introduce a new flag in ``cameras.xml``  a'la 
``<Crop x="0" y="0" width="0" height="0" read_from_exif="true"/>``
If cropping based on exif tags fails cropping shall fall back to values provided in  ``cameras.xml`` .

Reply:
Again, let's not call this "Exif crop". This is a pure kludge on Sony's part.

Reply:
> Well, at least it does in the DNG decoder, don't know if it'll just work out here.

As I said, it doesn't, see check here:
https://github.com/da-phil/rawspeed/blob/1972397b4a6091e3f13fd17a00860560bf15c827/src/librawspeed/decompressors/LJpegDecoder.cpp#L75

> Again, let's not call this "Exif crop". This is a pure kludge on Sony's part.

Is ``autocrop`` better? I don't think naming is our biggest problem with this LJPEG nightmare :sweat_smile:

Reply:
> As I said, it doesn't, see check here

Yeah, this is purely artificial, there is no reason for that constraint. The DNG decoder and its ljpeg decompressor work just fine without this constraint, as they should: https://github.com/darktable-org/rawspeed/blob/1ea45d926df9dbe904eb4b37f118277fbf6a16ab/src/librawspeed/decoders/DngDecoder.cpp#L289-L321

At this point, let's leave this PR as is if it's a working demo, and I'd let @LebedevRI decide on the refactoring going forward.

Reply:
As convoluted as the code is, ``DngDecoder`` eventually calls the same function (``LJpegDecoder::decode``)  in its own decompressor  ``AbstractDngDecompressor::decompressThread<7>()`` and additionally the ``LJpegDecompressor`` ctor does the same check [here](https://github.com/da-phil/rawspeed/blob/1972397b4a6091e3f13fd17a00860560bf15c827/src/librawspeed/decompressors/LJpegDecompressor.cpp#L76). Without those checks the code simply segfaults, as expected. For this to work we need to modify the code which runs over all tiles and allows to process the remaining fraction of tiles at the right and bottom borders. But yeah, this little "optimization" will not make a big difference. I'd rather continue with overall refactoring and creating a working solution for the M & S compressed files.

Reply:
> As convoluted as the code is, `DngDecoder` eventually calls the same function (`LJpegDecoder::decode`) in its own decompressor `AbstractDngDecompressor::decompressThread<7>()` and additionally the `LJpegDecompressor` ctor does the same check [here](https://github.com/da-phil/rawspeed/blob/1972397b4a6091e3f13fd17a00860560bf15c827/src/librawspeed/decompressors/LJpegDecompressor.cpp#L76). Without those checks the code simply segfaults, as expected. For this to work we need to modify the code which runs over all tiles and allows to process the remaining fraction of tiles at the right and bottom borders.

Again, we don't. In the DNG case, it really does work today (e.g. convert the uncompressed/lossy ILCE-7M4 to DNG: 7040x4688 is not a multiple of 256x256 tile size, and it doesn't throw). Before you even call `AbstractDngDecompressor::decompress()` and get to that check , the [DNG tile/slice dimensions are calculated for each one on creation](https://github.com/darktable-org/rawspeed/blob/1ea45d926df9dbe904eb4b37f118277fbf6a16ab/src/librawspeed/decoders/DngDecoder.cpp#L398), and [not constant/square on the last column/row](https://github.com/darktable-org/rawspeed/blob/1ea45d926df9dbe904eb4b37f118277fbf6a16ab/src/librawspeed/decompressors/AbstractDngDecompressor.h#L102-L108).

So, the correct behaviour (as outlined in the TIFF spec) is already implemented for the DNG case. Why Sony decided to make a mess of it, I'll never know...

As said, I'll let @LebedevRI decide how to refactor all of this best.

Reply:
Alright, due to the lack of understanding the rawspeed & LJPEG big picture yet I have to believe you ;)

Another question: after the changes in this PR we do not need the cropping which applies the values from `cameras.xml`, where is this applied and can we easily bypass it from within the ARW decoder?

Reply:
You probably just want to set `applyCrop` to false if you detect compression 7 [before this call](https://github.com/darktable-org/rawspeed/blob/1ea45d926df9dbe904eb4b37f118277fbf6a16ab/src/librawspeed/decoders/ArwDecoder.cpp#L365), but can't recall now if `decodeRawInternal()` comes before, or you'd have to detect it again here...

Reply:
Thanks, will have a look later, when I've got time to finish this work package for final review ;)

Reply:
Just checked all the rawspeed entrypoints and `decodeRaw()` always comes before `decodeMetaData()` (both calling their inherited decoder specific function), so this should work. Will give it a try now...

Reply:
by the way, by resetting the history on one of my test images using the `Exif.SubImage1.0x7038` crop I saw that this gives an area which still contains garbage from the tiling operation on the right and a black bar on the bottom
![Screenshot from 2023-06-06 22-53-40](https://github.com/darktable-org/rawspeed/assets/810285/2904c3aa-bc4f-4b71-821b-d5c6cf086a7c)

If I set the offset to 0,0 I only get valid pixels and no borders, so this way you can get the max. possible pixel area it seems. What's weird though is the size from `Exif.SubImage1.0x7038 = 9568, 6376`, but darktable crop module shows me 9567, 6376.
The DNG converted file shows 9504, 6336 in the crop module :sweat_smile: 

Then I did another test with way bigger crop to see where the garbage & borders begins and this proves that the values in `Exif.SubImage1.0x7038` seem to be quite accurate...
![Screenshot from 2023-06-06 23-51-22](https://github.com/darktable-org/rawspeed/assets/810285/87f8c40e-225e-4a58-8505-73dd94798e9a)

Reply:
Rather than resetting for validation, I'd remove the old images from the database (and their xmp sidecars), and import again.

The 0x7038 tag values, as mentioned, should result in the same size as the uncompressed/lossy case w/ `cameras.xml` crop. For 7RM5 this is, for uncompressed/lossy (9600-32)x(6376-0) which matches ljpeg compressed 0x7038 values 9568x6376.

Reply:
Agreed, then the current state (of cropping) in this PR makes sense, right?